[HTTP/OAS] Add Kibana OpenAPI bundling documentation (#195645)
elastic/kibana: oas_docs/README.md

### Step 2

At the second step, those domain-specific OpenAPI bundles are getting merged together by the `kbn-openapi-bundler` to produce the final resulting bundle.

### Scripts

The `oas_docs/scripts` folder contains scripts that point to the source domain-specific OpenAPI bundles and specify additional parameters for producing the final output bundle. Currently, there are two groups of scripts:

- The `merge_ess_oas.js` and `merge_serverless_oas.js` scripts produce production output bundles for ESS and Serverless respectively.

- The `merge_ess_oas_staging.js` and `merge_serverless_oas_staging.js` scripts produce staging output bundles for ESS and Serverless respectively.

### Output Kibana OpenAPI bundles

The `oas_docs/output` folder contains the final resulting Kibana OpenAPI bundles. There are two groups of them:

- `kibana.yaml` and `kibana.serverless.yaml` are production ready ESS and Serverless Kibana OpenAPI bundles to be published on the API reference documentation platform. Publishing process isn't fully automated yet and performed by the Docs Engineering team. These bundles are being deployed to https://www.elastic.co/docs/api.

> - `kibana.staging.yaml` and `kibana.serverless.staging.yaml` are staging ESS and Serverless Kibana OpenAPI bundles. Staging bundles currently aren't published anywhere and required for polishing API reference documentation on domains not yet included in the production ready Kibana OpenAPI bundles.

Author: We actively used `*staging*.yaml` files before including Security Solution's domain bundles in the main bundle. Though I agree we might not need them anymore.

In some cases engineers might need to have staging documentation so engineers and docs folks could review it before publishing it publicly. Currently it's possible to bundle a "custom" OpenAPI bundle locally by tinkering bundling configuration and publish it to a personal Bump.sh account for review. But it requires good understanding of the whole workflow and error prone without automation.

We could reiterate on that when we have a real use case.

- Use **Manually written** domain OpenAPI bundles if neither of the above doesn't work. It's considered as a temporary measure and highly discouraged as a long term approach.

### Step 2

At the second step domain OpenAPI bundles are merged together by `kbn-openapi-bundler` to produce the result bundle.

### Scripts

`scripts` folder contains scripts describing source domain OpenAPI bundles, result document configuration and output path. There are two groups currently

- `merge_ess_oas.js` and `merge_serverless_oas.js` to produce production ready Kibana OpenAPI bundles for ESS and Serverless respectively.

- `merge_ess_oas_staging.js` and `merge_serverless_oas_staging.js` to produce staging Kibana OpenAPI bundles for ESS and Serverless respectively. Staging bundles currently aren't published anywhere and required for polishing API reference documentation on domains not yet included in the production ready Kibana OpenAPI bundles.

### Output Kibana OpenAPI bundles

> `output` folder contains result Kibana OpenAPI bundles. There are two groups currently

Reviewer: ```suggestion
The `oas_docs/output` folder contains the final resulting Kibana OpenAPI bundles. There are two groups of them:
```

- `kibana.yaml` and `kibana.serverless.yaml` are production ready ESS and Serverless Kibana OpenAPI bundles to be published on the API reference documentation platform. Publishing process isn't fully automated yet and performed by the Docs Engineering team.

> - `kibana.staging.yaml` and `kibana.serverless.staging.yaml` are staging ESS and Serverless Kibana OpenAPI bundles. Staging bundles currently aren't published anywhere and required for polishing API reference documentation on domains not yet included in the production ready Kibana OpenAPI bundles.

Reviewer: I think we shouldn't repeat ourselves, the purpose of these bundles have been already described above in the Scripts section. Perhaps we could combine these sections into one to avoid duplication.

## Bundling commands

> Scripts in the `scripts` folder bundle individual Kibana OpenAPI bundles. There is a makefile to simplify the workflow. The following makefile targets are available

Reviewer: ```suggestion
Besides the scripts in the `oas_docs/scripts` folder, there is an `oas_docs/makefile` to simplify the workflow. The following makefile targets are available:
```

| `api-docs-lint-errs`       | Lints built result bundles for errors                                                                                                                                                                              |
| `api-docs-preview`         | Generates (ESS + Serverless) Kibana OpenAPI bundles preview                                                                                                                                                        |
| `api-docs-overlay`         | Applies [overlays](https://docs.bump.sh/help/specification-support/overlays/) from `overlays` folder to the Kibana OpenAPI bundles and generate `*.new.yaml` files. Overlays help to fine tune the result bundles. |
| `api-docs-overlay-preview` | Generates a preview for bundles produced by `api-docs-overlay`                                                                                                                                                     |

## Bundling automation

> To have always up to date Kibana OpenAPI bundles bundling runs as a part of `Checks` step in CI on every PR build and on merge builds. If there are any changes in source OpenAPI files these changes get to the updated Kibana OpenAPI bundles and CI commits the changes. In that case CI is marked as failed and needs to be restarted.

Reviewer: ```suggestion
To keep the final resulting bundles always up-to-date and in sync with the source OpenAPI specs, the bundling runs as part of the `Checks` step in CI on every PR build and on merge builds. If there are any changes to the source OpenAPI files, these changes get propagated to the final bundles and CI commits the changes. In that case the build is marked as failed and needs to be restarted.
```

Do we really push commits from CI in this case, or just fail without committing?

## Bundling solution domains

> ### Security Solution

Reviewer: Besides paths to the bundling scripts and output folders, can we please elaborate on how these scripts are integrated into the CI?

Author: I've added a new sub section but it basically very similar to `Bundling automation` section.

#### Security AI Assistant

Root folder `x-pack/packages/kbn-elastic-assistant-common`.

|                             |                                                                                |
| --------------------------- | ------------------------------------------------------------------------------ |
| Source OpenAPI files folder | `impl`                                                                         |
| Bundling script             | `scripts/openapi/bundle.js`                                                    |
| Result Serverless bundle    | `docs/openapi/serverless/elastic_assistant_api_2023_10_31.bundled.schema.yaml` |

> | Result ESS bundle           | `docs/openapi/ess/elastic_assistant_api_2023_10_31.bundled.schema.yaml`        |

Reviewer: There's a lot of tables here and perhaps too much details - some of this information can become outdated in the future. Let's try to make it a bit more succinct and resilient to refactorings. What do you think about a simple unordered list:

- Security AI Assistant API
  - Bundling script: `x-pack/packages/kbn-elastic-assistant-common/scripts/openapi/bundle.js`
  - Bundles' location: `x-pack/packages/kbn-elastic-assistant-common/docs/openapi/{ess|serverless}`
- Security Detections API
  - Bundling script: `x-pack/plugins/security_solution/scripts/openapi/bundle_detections.js`
  - Bundles' location: `x-pack/plugins/security_solution/docs/openapi/{ess|serverless}`
- etc
